### PR TITLE
Fix torchax tensor flatten bug on empty tensors

### PR DESCRIPTION
**Pull request by cychiuak**
Flatten on tensor of shape contains zero is supported in PyTorch, but not through torchax.
```
import torch
import torchax

with torchax.default_env():
    value = torch.ones((16, 0))
    print(value.flatten(0, -2).shape) # works in CPU > torch.Size([16, 0])
    value = value.to(device='jax')
    print(value.flatten(0, -2)) # throws exception
```
This is due to jax.numpy fails to calculate the correct dimension when input dimension contains 0 and -1 at the same time. It returns a division by zero error.
```
import jax.numpy as jnp

print(jnp.reshape(jnp.ones((2, 3)), (-1, 3)).shape) # works > (2, 3)
value = jnp.ones((16, 0))
print(jnp.reshape(value, (16, 0)).shape) # works > (16, 0)
print(jnp.reshape(value, (-1, 0)).shape) # should work but throws exception
```
The new fix avoids sending -1 into jnp.shape by explicitly calculating the flattening dimension to to fix this error:
```
flattened_size = math.prod(self._elem.shape[start_dim : end_dim + 1])
new_shape = self._elem.shape[:start_dim] + (flattened_size,) + self._elem.shape[end_dim + 1 :]
```

### CODE REVIEW

File: test/test_functions.py [hunk 1]
```diff
@@ -80,10 +80,24 @@ def test_bernoulli_inplace(self):
...
       self.assertEqual(tuple(a.shape), (6, 4))
 
+      a = torch.randn(()).to(device='jax')
+      a = a.flatten(0, -1)
+      self.assertEqual(tuple(a.shape), (1,))
+
+      # New test case for 0-sized tensor with negative end_dim
+      a = torch.ones((16, 0)).to(device='jax')
+      a = a.flatten(0, -2)
+      self.assertEqual(tuple(a.shape), (16, 0))
```
Comment (wdhongtw) [inline test/test_functions.py]: nit: maybe self.assertEqual(a.shape, torch.Size([16, 0]))

File: test/test_functions.py [hunk 2]
```diff
@@ -80,10 +80,24 @@ def test_bernoulli_inplace(self):
...
   def test_flatten(self):
     with self.env:
-      a = torch.randn((2, 3, 4))
+      a = torch.randn((2, 3, 4)).to(device='jax')
       a = a.flatten(0, 1)
       self.assertEqual(tuple(a.shape), (6, 4))
 
+      a = torch.randn(()).to(device='jax')
+      a = a.flatten(0, -1)
+      self.assertEqual(tuple(a.shape), (1,))
```
Comment (wdhongtw) [inline test/test_functions.py]: This is correct from the aspect of math, but still quite confusing.
I will suggest not including this test case.

File: test/test_functions.py [hunk 3]
```diff
@@ -80,10 +80,24 @@ def test_bernoulli_inplace(self):
...
+      # New test case for 0-sized tensor with negative end_dim
+      a = torch.ones((16, 0)).to(device='jax')
+      a = a.flatten(0, -2)
+      self.assertEqual(tuple(a.shape), (16, 0))
+
+      # Another test cases for empty dimensions
+      a = torch.randn((2, 0, 0, 5)).to(device='jax')
+      a = a.flatten(2, 3)
+      self.assertEqual(tuple(a.shape), (2, 0, 0))
+
```
Comment (wdhongtw) [inline test/test_functions.py:106]: Probably add more test cases.

**Review by wdhongtw — CHANGES_REQUESTED** [review]
LGTM! 👍

Just a few suggestions.

**Review by wdhongtw — APPROVED** [review]
